### PR TITLE
feat(bucket-brigade): non-PPO best-response improvability oracle (AC#1 gate)

### LINKED ISSUE #259: PPO best-response improves critic fit (EV≈0.57) but not team return on bucket-brigade — diagnose the policy-update failure (with improvability gate)
## Summary

The PPO best-response (BR) on the bucket-brigade no-convergence cells **does not improve team return**, and the 2026-06-28 cluster probe sweep (PR #256) **moved the diagnosis**: the critic now fits well, yet the policy still doesn't improve. This is the load-bearing blocker under the #134 negative result and gates #230.

This supersedes the framing of the now-closed #239 (PR #248) and #252. Those said *"the critic doesn't fit."* That is **no longer the symptom.**

## New evidence (refutes the critic-fit framing)

`train_br_probe`, **8192 rollout**, 3 cells × 5 levers, NdArray CPU (full data: `docs/research/2026-06-bucket-brigade-validation.md`, "Stage-1 cluster probe sweep"):

| lever (β=0.5) | critic `ev` early→late | `mean_ep_return` early5→late5 |
|------|------|------|
| `base` 8192×8 | 0.00 → **0.565** | −27 884 → −27 285 (flat) |
| `vf1` VF_COEF=1.0 | 0.00 → **0.583** | −27 455 → −28 454 (flat) |
| `br16` 16 trainsteps | 0.00 → **0.575** | −27 727 → −27 302 (flat) |
| `bigroll` 16384 rollout | 0.00 → 0.328 | −28 046 → −27 983 (flat) |
| `rs01` BR_REWARD_SCALE=0.01 | 0.00 → 0.053 | −28 076 → −27 303 (flat) |

At 8192 the critic reaches **EV ≈ 0.57** — a genuinely well-fit value head, past the prior ~0.48 "ceiling" — and `mean_ep_return` is **dead flat** across every lever and all 3 cells. **0 of 15 evaluable runs raised team return.** Critic fit is conclusively *not* the bottleneck.

The open question is therefore: **why does the PPO best-response policy not improve team return even when the advantage signal is well-estimated?**

## ⚠️ Acceptance criterion #1 — improvability gate (do this FIRST; do NOT patch before it)

Before touching the PPO update, **establish whether a better-than-uniform BR even exists** on at least one cell. The flat return may be a property of the game/reward (BR vs frozen uniform opponents genuinely near-flat), in which case **no fix exists and this research direction is exhausted** — that is a valid, documentable outcome, not a bug.

- Compute a **return ceiling** on ≥1 cell via a non-PPO method: brute-force / CEM / scripted-specialist search for the single BR agent against the frozen uniform opponents `train_br_probe` uses, reporting best achievable per-step **team** return vs the uniform baseline.
- **If the gap is negligible** (oracle ≈ uniform): STOP. Document "BR-vs-uniform return is near-flat on cell X; the #134 direction is exhausted at the BR level," update #134/#230, and close. No policy fix is warranted.
- **If a real gap exists**: proceed to AC#2 with the measured ceiling as the target.

## Acceptance criterion #2 — diagnose the policy-update failure (gated on AC#1 showing a real gap)

With a concrete improvable gap in hand, instrument and explain why PPO doesn't close it. Candidate mechanisms to rule in/out (instrument, don't guess):

- **Credit assignment / reward channel**: confirm the BR optimizes the same quantity `mean_ep_return` measures. Shared/team reward may give the BR agent a marginal contribution too small to drive its own policy gradient; check the per-agent reward the BR actually trains on vs the logged team return.
- **Advantage → update signal**: log advantage mean/std post-normalization, PPO ratio/clip-fraction, per-factor (`[house, mode, signal]`) policy-gradient norms and entropy. A well-fit critic with near-zero *effective* advantage signal points here.
- **Exploration / entropy floor** keeping the policy near uniform; **LR**; **factored-action** policy-gradient composition over the 3 heads.
- **Deliverable**: a `train_br_probe` config that moves `mean_ep_return` measurably toward the AC#1 ceiling on that cell, with the mechanism identified.

## Reproduction

```bash
CELL=beta05 ITERATIONS=40 ROLLOUT_STEPS=8192 \
  cargo run --release --example train_br_probe --features "training,env-bucket-brigade"
# watch br[... ev=...] vs mean_ep_return — EV climbs to ~0.57, return stays ~−28k
```

## Relevant files

- `examples/games/bucket_brigade/train_br_probe.rs` — the fast single-BR A/B harness (#241)
- `src/multi_agent/joint.rs` — the joint PPO update (advantage, EV, clip)
- `src/env/games/bucket_brigade/` — env / reward structure (for AC#1 oracle + credit-assignment check)
- `src/multi_agent/bucket_brigade_metrics.rs`, `bucket_brigade_baselines.rs` — `gap_closed_cell`, cell baselines

## Lineage

Closed predecessors: #239 (critic-fit root cause + PR #248 fix), #252 (fix under-delivers at scale), #241 (EV logging + probe harness), #235 (rollout batching). This issue is the **next link**: the #248 fix worked (critic fits at scale now) but revealed the real blocker is the policy update, not the critic.

`tier:goal-advancing` — the actual unblocker for #134 / #230.

### PR DESCRIPTION
**Pull request by rjwalters**
## Summary

Implements the **improvability gate (AC#1)** for issue #259: before touching the PPO/policy update, establish — via a **non-PPO** method — whether a better-than-uniform best-response (BR) even *exists* against the same frozen-uniform opponents `train_br_probe` trains against.

It does, decisively: **no.** This is **outcome 1** of the issue (oracle ≈ uniform → flat). The deliverable is therefore the **oracle harness + documented negative result**, and AC#2 (instrument/patch the PPO update) is **correctly skipped** because its precondition (a real improvable gap) is not met. **No PPO code is touched** — per the issue's explicit hard ordering.

## AC#1 oracle finding (the numbers)

Scripted (NN-free, PPO-free) policies for the single BR agent vs 3 frozen uniform-random opponents, 400 eval episodes/cell, canonical cell β=0.5 (all three cells identical — see note):

| BR policy (agent 0; other 3 uniform) | per-step **team** | per-step **BR-agent** | per-ep **BR-agent** |
|------|------|------|------|
| `uniform` (baseline) | −673.950 | −205.448 | −27 634 |
| `always_rest` | −673.886 | −201.979 | −27 372 |
| `specialist` (gap_closed=1.0 ref) | −678.519 | −196.504 | −27 831 |
| `firefighter[owned, work=1.0]` | −675.822 | −194.403 | −27 144 |
| `firefighter[any, work=1.0]` | −674.063 | −204.604 | −29 732 |
| `search_best firefighter[owned, work=0.853]` | −676.583 | −197.760 | −27 587 |

- **Ceiling = `always_rest` at −673.886 vs baseline −673.950 → team gap = +0.064/step = +0.01% of |baseline|. Flat.**
- The strongest *known* policy (specialist) and an aggressive any-house firefighter are **worse** on team return than doing nothing — fighting fires costs `c=0.5`/night while 3 uniform teammates let the village ruin regardless.
- Even the BR agent's **own** return — the quantity `train_br_probe` actually optimizes (directly addressing the AC#2 credit-assignment hypothesis) — has a ceiling of only **+0.95%** above uniform, and that edge comes from *resting more* (avoiding wasted work cost), not from coordinating a save.

This is the BR-level confirmation of the #134 negative result: with 3 of 4 agents frozen uniform, the BR's marginal contribution to team return is ~0, so a well-fit critic correctly reports near-zero advantage and the policy correctly stays near uniform. Reviving #134 requires changing the **game/credit-assignment regime**, not the PPO optimizer. #230 stays correctly gated.

## Which outcome branch + why

**Branch 1 (flat).** The measured ceiling is statistically indistinguishable from uniform on team return (+0.01%) and near-flat on the BR's own return (+0.95%). Per the issue, the correct deliverable is the oracle + documented finding, not a speculative policy patch.

## Changes

- `src/multi_agent/bucket_brigade_oracle.rs` — reusable, unit-tested library. Scores `uniform`/`always_rest`/`specialist`/firefighter battery + a 64-sample random search over a `FirefighterParams { scope_owned_only, work_prob }` family that *contains* the specialist endpoint. Reuses the existing env and the same frozen-uniform-opponent setup `train_br_probe` constructs (true-uniform idealization of its freshly-init ≈uniform nets). Shared per-episode seed stream for variance reduction.
- `examples/games/bucket_brigade/br_oracle.rs` — thin CLI wrapper (`CELL`, `EVAL_EPISODES`, `SEARCH_EPISODES`, `NUM_SEARCH`, `SEED`, `STEP_CAP`).
- `src/multi_agent/mod.rs`, `Cargo.toml` — module + example registration (gated `env-bucket-brigade` / `training,env-bucket-brigade`, matching siblings).
- `docs/research/2026-06-bucket-brigade-validation.md` — appended dated section with the finding, table, implication for #134/#230, and repro command.

### Note on the across-cell identity

All three cells (β=0.1/0.5/0.9) produced identical aggregate statistics under the shared seed stream. This matches the repo's own committed per-cell baselines (`MINSPEC_RANDOM_BETA01/05/09` and `MINSPEC_SPECIALIST_BETA01/05/09` are byte-identical in `bucket_brigade_metrics.rs`) and the documented "wasteland-collapse" dynamics. The policy comparison — the actual AC#1 question — is fully live and differentiated (specialist −678.5 vs uniform −673.9 vs always_rest −673.9).

## Acceptance Criteria Verification

| Criterion | Status | Verification |
|-----------|--------|--------------|
| AC#1 oracle implemented AND run; numeric ceiling vs uniform reported | ✅ | `br_oracle` example + library; numbers above; repro in PR/doc |
| Non-PPO method (brute-force/scripted-specialist) on ≥1 cell | ✅ | Scripted battery + random search; ran all 3 cells |
| Reports best per-step **team** return vs uniform baseline | ✅ | Ceiling −673.886 vs −673.950 = +0.01% |
| Branch 1 (flat): document finding in validation doc, note #134/#230 implication, skip AC#2 explicitly | ✅ | Dated section appended; AC#2 skip justified here and in doc |
| No speculative PPO/policy patch | ✅ | `git diff` touches no PPO/`joint.rs` code |
| Repo compiles; tests pass | ✅ | `cargo build` (default) + `cargo build --release --example br_oracle`; 2 lib tests pass; clippy clean |

## Test Plan

- `cargo test --release --lib --features "training,env-bucket-brigade" bucket_brigade_oracle` → 2 passed (oracle runs/sane; firefighter≡specialist on burning-owned house).
- `cargo build` (default features) and `cargo build --release --example br_oracle --features "training,env-bucket-brigade"` → clean; `cargo clippy --lib` clean.
- Ran the oracle on all three cells (numbers above).

Repro:
```bash
for c in beta01 beta05 beta09; do
  CELL=$c EVAL_EPISODES=400 \
    cargo run --release --example br_oracle --features "training,env-bucket-brigade"
done
```

Closes #259

### CODE REVIEW

**Comment by rjwalters** [comment]
## Verdict: Approved (loom:pr)

Verified AC#1 against the actual diff, not the summary. This is a genuine, well-scoped improvability-gate deliverable.

### AC#1 oracle is real and methodologically sound
- `src/multi_agent/bucket_brigade_oracle.rs` implements an honest non-PPO oracle: NN-free / Burn-free / PPO-free scripted policy battery (`uniform`, `always_rest`, `specialist`, owned/any firefighters) plus a 64-sample random search over the `FirefighterParams { scope_owned_only, work_prob }` family. The family provably *contains* the `specialist_action` (gap_closed==1.0) endpoint — unit-tested in `firefighter_owned_matches_specialist_on_burning_owned`.
- It uses the **same** single-BR-vs-frozen-uniform-opponents setup `train_br_probe` uses (`BR_AGENT = 0`, N-1 = 3 uniform opponents), with the uniform-opponent idealization of the freshly-init nets honestly disclosed in the module docs.
- It **actually compares distinct candidate policies**, not just re-measures uniform: specialist -678.519, firefighter[owned] -675.822, always_rest -673.886 are all materially different from uniform -673.950. The AC#1 question is live and differentiated.
- The numeric ceiling is reported in both PR body and `docs/research/2026-06-bucket-brigade-validation.md`: ceiling = always_rest -673.886 vs baseline -673.950 = **+0.064/step (+0.01% of |baseline|)** team return; BR-agent own-return ceiling only +0.95% (achieved by resting, not saving). Flat.

### Scope ordering respected
- Flat oracle correctly triggered **Branch 1**: document the exhausted #134 BR-level direction, keep #230 gated, and **explicitly skip AC#2** as precondition-unmet. Confirmed the diff touches **no PPO code** — only Cargo.toml, the new oracle lib + example, mod.rs registration, and the validation doc. `joint.rs` is untouched. No speculative policy patch.

### Quality
- CI fully green (Clippy, Rustfmt, Docs, Tests Linux+macOS NdArray, Security Audit all pass); mergeStateStatus CLEAN.
- 2 focused unit tests (oracle-runs-and-sane + firefighter==specialist equivalence). Diff is tight and well-documented.
- The across-cell byte-identity caveat is honest and corroborated: the repo's own committed `MINSPEC_RANDOM_BETA01/05/09` (-605.5118) and `MINSPEC_SPECIALIST_BETA01/05/09` (-602.0938) are likewise byte-identical, so this reflects a pre-existing wasteland-collapse property of the env, not a bug introduced here. The within-cell policy comparison — the actual AC#1 question — is fully live, so the caveat does not undermine the flat conclusion.

LGTM.
